File: server/routes/handwriting.js
/**
 * Handwriting Routes
 * Handles GAN-style handwritten date generation
 */

import express from 'express';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs';
import { v4 as uuidv4 } from 'uuid';
import { db, logAuditAction } from '../database/db.js';
import handwritingGenerator from '../services/handwriting-generator.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const router = express.Router();

// Ensure output directory exists
const outputDir = path.join(__dirname, '..', 'uploads', 'generated');
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

// GET /api/handwriting/styles - Get available handwriting styles
router.get('/styles', (req, res) => {
  try {
    const styles = db.prepare('SELECT * FROM handwriting_styles WHERE is_active = 1').all();
    res.json(styles);
  } catch (error) {
    console.error('Error fetching styles:', error);
    res.status(500).json({ error: error.message });
  }
});

// POST /api/handwriting/generate-date - Generate handwritten date image
router.post('/generate-date', async (req, res) => {
  try {
    const { 
      date, 
      style = 'Natural Cursive', 
      format = 'full',
      width = 300,
      height = 80
    } = req.body;

    const dateObj = date ? new Date(date) : new Date();

    // Get style configuration
    const styleConfig = db.prepare('SELECT * FROM handwriting_styles WHERE name = ?').get(style);

    const result = await handwritingGenerator.generateHandwrittenDate({
      date: dateObj,
      format,
      width: parseInt(width),
      height: parseInt(height),
      style: styleConfig || undefined
    });

    // Save to file
    const filename = `date_${uuidv4()}.png`;
    const filePath = path.join(outputDir, filename);
    
    // Extract base64 data and save
    const base64Data = result.image.replace(/^data:image\/\w+;base64,/, '');
    fs.writeFileSync(filePath, Buffer.from(base64Data, 'base64'));

    logAuditAction(null, 'System', 'DATE_GENERATED', 'handwriting', null, { date: dateObj.toISOString(), style });

    res.json({
      success: true,
      image: result.image,
      filePath: `/uploads/generated/${filename}`,
      metadata: {
        date: dateObj.toISOString(),
        style,
        format,
        dimensions: { width: parseInt(width), height: parseInt(height) }
      }
    });
  } catch (error) {
    console.error('Error generating date:', error);
    res.status(500).json({ error: error.message, success: false });
  }
});

// POST /api/handwriting/generate-variations - Generate multiple date variations
router.post('/generate-variations', async (req, res) => {
  try {
    const { date, count = 5 } = req.body;
    const dateObj = date ? new Date(date) : new Date();

    const variations = await handwritingGenerator.generateDateVariations(dateObj, Math.min(count, 10));

    res.json({
      success: true,
      variations,
      date: dateObj.toISOString()
    });
  } catch (error) {
    console.error('Error generating variations:', error);
    res.status(500).json({ error: error.message, success: false });
  }
});

// POST /api/handwriting/composite - Composite signature with handwritten date
router.post('/composite', async (req, res) => {
  try {
    const { 
      signatureId,
      signaturePath,
      date,
      style = 'Natural Cursive',
      position = 'below',
      outputWidth = 400,
      outputHeight = 150
    } = req.body;

    let sigPath;
    
    if (signatureId) {
      const signature = db.prepare('SELECT file_path FROM signature_assets WHERE id = ? AND status = "Approved"').get(signatureId);
      if (!signature) {
        return res.status(404).json({ error: 'Approved signature not found' });
      }
      sigPath = path.join(__dirname, '..', signature.file_path);
    } else if (signaturePath) {
      sigPath = signaturePath.startsWith('/') 
        ? path.join(__dirname, '..', signaturePath)
        : signaturePath;
    } else {
      return res.status(400).json({ error: 'signatureId or signaturePath is required' });
    }

    if (!fs.existsSync(sigPath)) {
      return res.status(404).json({ error: 'Signature file not found' });
    }

    const dateObj = date ? new Date(date) : new Date();

    const result = await handwritingGenerator.compositeSignatureWithDate(
      sigPath,
      dateObj,
      {
        style,
        position,
        outputWidth: parseInt(outputWidth),
        outputHeight: parseInt(outputHeight)
      }
    );

    // Save composite image
    const filename = `composite_${uuidv4()}.png`;
    const filePath = path.join(outputDir, filename);
    
    const base64Data = result.image.replace(/^data:image\/\w+;base64,/, '');
    fs.writeFileSync(filePath, Buffer.from(base64Data, 'base64'));

    logAuditAction(null, 'System', 'SIGNATURE_COMPOSITE_CREATED', 'handwriting', null, { signatureId, date: dateObj.toISOString() });

    res.json({
      success: true,
      image: result.image,
      filePath: `/uploads/generated/${filename}`,
      metadata: result.metadata
    });
  } catch (error) {
    console.error('Error creating composite:', error);
    res.status(500).json({ error: error.message, success: false });
  }
});

// POST /api/handwriting/styles - Create new handwriting style
router.post('/styles', (req, res) => {
  try {
    const {
      name,
      fontFamily,
      baseSize = 24,
      rotationVariance = 2.0,
      spacingVariance = 1.5,
      strokeWidthVariance = 0.5
    } = req.body;

    if (!name || !fontFamily) {
      return res.status(400).json({ error: 'Name and fontFamily are required' });
    }

    const stmt = db.prepare(`
      INSERT INTO handwriting_styles (name, font_family, base_size, rotation_variance, spacing_variance, stroke_width_variance)
      VALUES (?, ?, ?, ?, ?, ?)
    `);
    
    const result = stmt.run(name, fontFamily, baseSize, rotationVariance, spacingVariance, strokeWidthVariance);

    const newStyle = db.prepare('SELECT * FROM handwriting_styles WHERE id = ?').get(result.lastInsertRowid);
    res.status(201).json(newStyle);
  } catch (error) {
    console.error('Error creating style:', error);
    res.status(500).json({ error: error.message });
  }
});

// DELETE /api/handwriting/styles/:id - Delete handwriting style
router.delete('/styles/:id', (req, res) => {
  try {
    db.prepare('UPDATE handwriting_styles SET is_active = 0 WHERE id = ?').run(req.params.id);
    res.json({ message: 'Style deactivated' });
  } catch (error) {
    console.error('Error deleting style:', error);
    res.status(500).json({ error: error.message });
  }
});

// GET /api/handwriting/preview - Preview handwritten date without saving
router.get('/preview', async (req, res) => {
  try {
    const { date, style = 'Natural Cursive', format = 'full' } = req.query;
    const dateObj = date ? new Date(date) : new Date();

    const result = await handwritingGenerator.generateHandwrittenDate({
      date: dateObj,
      format,
      width: 300,
      height: 80
    });

    res.json({
      success: true,
      image: result.image,
      date: dateObj.toISOString()
    });
  } catch (error) {
    console.error('Error previewing date:', error);
    res.status(500).json({ error: error.message, success: false });
  }
});

export { router as handwritingRoutes };

File: backend/app/routers/handwriting.py
"""
Handwriting Routes - PyTorch GAN handwriting generation
"""
from fastapi import APIRouter, Request, HTTPException, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from typing import Optional
from pydantic import BaseModel
from datetime import datetime
import uuid
from pathlib import Path

from app.database import get_db, HandwritingStyle, SignatureAsset
from app.utils.timezone import get_ph_now, format_ph_datetime

router = APIRouter()
OUTPUT_DIR = Path(__file__).parent.parent.parent / "uploads" / "generated"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


class DateGenerateRequest(BaseModel):
    date: Optional[str] = None
    style: str = "Natural Cursive"
    format: str = "full"
    width: int = 300
    height: int = 80


class CompositeRequest(BaseModel):
    signatureId: Optional[int] = None
    signaturePath: Optional[str] = None
    date: Optional[str] = None
    style: str = "Natural Cursive"
    outputWidth: int = 400
    outputHeight: int = 150


@router.get("/styles")
async def list_styles(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(HandwritingStyle).where(HandwritingStyle.is_active == True))
    return result.scalars().all()


@router.post("/generate-date")
async def generate_handwritten_date(request: Request, data: DateGenerateRequest):
    model = request.app.state.handwriting_model
    if not model:
        raise HTTPException(status_code=500, detail="Handwriting model not initialized")

    date_obj = datetime.fromisoformat(data.date) if data.date else get_ph_now()

    result = await model.generate_date(
        date=date_obj,
        style=data.style,
        format_type=data.format,
        width=data.width,
        height=data.height
    )

    filename = f"date_{uuid.uuid4()}.png"
    file_path = OUTPUT_DIR / filename

    if result.get("image_bytes"):
        with open(file_path, "wb") as f:
            f.write(result["image_bytes"])

    return {
        "success": True,
        "image": result.get("image_base64"),
        "filePath": f"/uploads/generated/{filename}",
        "metadata": {
            "date": date_obj.isoformat(),
            "style": data.style,
            "format": data.format
        }
    }


@router.post("/generate-variations")
async def generate_variations(request: Request, date: Optional[str] = None, count: int = 5):
    model = request.app.state.handwriting_model
    if not model:
        raise HTTPException(status_code=500, detail="Handwriting model not initialized")

    date_obj = datetime.fromisoformat(date) if date else get_ph_now()
    variations = await model.generate_variations(date_obj, min(count, 10))
    return {"success": True, "variations": variations, "date": date_obj.isoformat()}


@router.post("/composite")
async def composite_signature_date(request: Request, data: CompositeRequest, db: AsyncSession = Depends(get_db)):
    model = request.app.state.handwriting_model
    if not model:
        raise HTTPException(status_code=500, detail="Handwriting model not initialized")

    sig_path = None
    if data.signatureId:
        result = await db.execute(
            select(SignatureAsset)
            .where(SignatureAsset.id == data.signatureId, SignatureAsset.status == "Approved")
        )
        sig = result.scalar_one_or_none()
        if not sig:
            raise HTTPException(status_code=404, detail="Approved signature not found")
        sig_path = str(Path(__file__).parent.parent.parent / sig.file_path.lstrip("/"))
    elif data.signaturePath:
        sig_path = data.signaturePath
    else:
        raise HTTPException(status_code=400, detail="signatureId or signaturePath required")

    date_obj = datetime.fromisoformat(data.date) if data.date else get_ph_now()

    result = await model.composite_signature_with_date(
        signature_path=sig_path,
        date=date_obj,
        style=data.style,
        output_width=data.outputWidth,
        output_height=data.outputHeight
    )

    filename = f"composite_{uuid.uuid4()}.png"
    return {
        "success": True,
        "image": result.get("image_base64"),
        "filePath": f"/uploads/generated/{filename}"
    }


@router.get("/preview")
async def preview_date(request: Request, date: Optional[str] = None, style: str = "Natural Cursive"):
    model = request.app.state.handwriting_model
    if not model:
        raise HTTPException(status_code=500, detail="Handwriting model not initialized")

    date_obj = datetime.fromisoformat(date) if date else get_ph_now()
    result = await model.generate_date(date=date_obj, style=style)
    return {"success": True, "image": result.get("image_base64"), "date": date_obj.isoformat()}


# =============================================================================
# Style Extraction & Custom Font Endpoints
# =============================================================================

class ExtractStyleRequest(BaseModel):
    signatureId: int


class WeekdayDatesRequest(BaseModel):
    signatureId: Optional[int] = None
    weekOffset: int = 0


class CustomCompositeRequest(BaseModel):
    signatureId: int
    date: Optional[str] = None
    outputWidth: int = 400
    outputHeight: int = 150


@router.post("/extract-style")
async def extract_signature_style(
    request: Request,
    data: ExtractStyleRequest,
    db: AsyncSession = Depends(get_db)
):
    """Extract style characteristics from a signature for matching date generation."""
    model = request.app.state.handwriting_model
    if not model:
        raise HTTPException(status_code=500, detail="Handwriting model not initialized")

    # Get signature from database
    result = await db.execute(
        select(SignatureAsset).where(SignatureAsset.id == data.signatureId)
    )
    sig = result.scalar_one_or_none()
    if not sig:
        raise HTTPException(status_code=404, detail="Signature not found")

    sig_path = str(Path(__file__).parent.parent.parent / sig.file_path.lstrip("/"))

    style_info = await model.extract_signature_style(sig_path)

    return {
        "success": True,
        "signatureId": data.signatureId,
        "styleVector": style_info.get("style_vector"),
        "styleParams": style_info.get("style_params")
    }


@router.post("/weekday-dates")
async def generate_weekday_dates(
    request: Request,
    data: WeekdayDatesRequest,
    db: AsyncSession = Depends(get_db)
):
    """Generate dated signatures for a week (Monday-Friday) matching signature style."""
    model = request.app.state.handwriting_model
    if not model:
        raise HTTPException(status_code=500, detail="Handwriting model not initialized")

    sig_path = None
    if data.signatureId:
        result = await db.execute(
            select(SignatureAsset).where(SignatureAsset.id == data.signatureId)
        )
        sig = result.scalar_one_or_none()
        if sig:
            sig_path = str(Path(__file__).parent.parent.parent / sig.file_path.lstrip("/"))

    dates = await model.generate_weekday_dates(
        signature_path=sig_path,
        week_offset=data.weekOffset
    )

    return {
        "success": True,
        "weekOffset": data.weekOffset,
        "signatureId": data.signatureId,
        "dates": dates
    }


@router.post("/custom-composite")
async def create_custom_composite(
    request: Request,
    data: CustomCompositeRequest,
    db: AsyncSession = Depends(get_db)
):
    """Create composite signature with date using custom font matching signature style."""
    model = request.app.state.handwriting_model
    if not model:
        raise HTTPException(status_code=500, detail="Handwriting model not initialized")

    # Get approved signature
    result = await db.execute(
        select(SignatureAsset)
        .where(SignatureAsset.id == data.signatureId, SignatureAsset.status == "Approved")
    )
    sig = result.scalar_one_or_none()
    if not sig:
        raise HTTPException(status_code=404, detail="Approved signature not found")

    sig_path = str(Path(__file__).parent.parent.parent / sig.file_path.lstrip("/"))

    date_obj = datetime.fromisoformat(data.date) if data.date else get_ph_now()

    composite_result = await model.composite_signature_with_custom_date(
        signature_path=sig_path,
        date=date_obj,
        output_width=data.outputWidth,
        output_height=data.outputHeight
    )

    # Save composite image
    filename = f"custom_composite_{uuid.uuid4()}.png"
    file_path = OUTPUT_DIR / filename

    if composite_result.get("image_bytes"):
        with open(file_path, "wb") as f:
            f.write(composite_result["image_bytes"])

    return {
        "success": True,
        "image": composite_result.get("image_base64"),
        "filePath": f"/uploads/generated/{filename}",
        "metadata": composite_result.get("metadata")
    }


@router.get("/custom-font-status")
async def check_custom_font_status(request: Request):
    """Check if custom font images are loaded."""
    model = request.app.state.handwriting_model
    if not model:
        return {"loaded": False, "fontCount": 0}

    font_count = len(model.custom_font_renderer.font_images) if hasattr(model, 'custom_font_renderer') else 0

    return {
        "loaded": font_count > 0,
        "fontCount": font_count,
        "availableChars": list(model.custom_font_renderer.font_images.keys()) if font_count > 0 else []
    }

